pkg/util/deeptest/statictesthelper.go

```diff
@@ -0,0 +1,142 @@
+// Copyright 2024 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package deeptest
+
+import (
+	"fmt"
+	"reflect"
+	"regexp"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+)
+
+type staticTestHelper struct {
+	ignorePath []regexp.Regexp
+}
+
+// NewStaticTestHelper creates a new helper to help testing the deep equality of two objects.
+func NewStaticTestHelper(ignorePath []string) *staticTestHelper {
+	pathRegexp := make([]regexp.Regexp, 0, len(ignorePath))
+	for _, path := range ignorePath {
+		pathRegexp = append(pathRegexp, *regexp.MustCompile(path))
+	}
+	return &staticTestHelper{ignorePath: pathRegexp}
+}
+
+func (h *staticTestHelper) shouldIgnore(path string) bool {
+	for _, reg := range h.ignorePath {
+		if reg.MatchString(path) {
+			return true
+		}
+	}
+	return false
+}
+
+// AssertNotEmpty asserts that the value and its every fields are not empty.
+func (h *staticTestHelper) AssertNotEmpty(t *testing.T, valObj reflect.Value, path string) {
```

Comment:
How about to introduce a method `AssertEveryFieldNotEqual(t *testing.T, val1, val2 reflect.Value, path)` ? For example, `NewStaticExprContext()` does not return an object with the "go default fields", it returns "variable default fields" with some fields set to `variable.DefaultXXX`.

Reply:
Good. I've changed to `AssertEveryFieldNotEqual`, and it actually found one issue :facepalm: